May 2018 workshop

CZagrobelny opened the pull request:
Branched off of Railsbridge master (not Railsbridge NYC).
This includes:
1. Docker changes for installfest
2. Docker changes for intro to rails
3. Docker changes to Job board
4. Curriculum changes in intro to rails (MVC and adding hashes to the intro ruby section)

To view changes live: https://serene-thicket-60610.herokuapp.com

CZagrobelny committed:
add a section on hashes

CZagrobelny committed:
cleanup and restore else explanation

CZagrobelny committed:
add small group details for practice problem

CZagrobelny committed:
remove link to railsbridge instance explanation

CZagrobelny committed:
restore reminder to exit irb

CZagrobelny committed:
fix syntax error

CZagrobelny committed:
remove extraneous new line

CZagrobelny committed:
eliminate unnecessary new line

CZagrobelny committed:
add request cycle and more detailed explanation of mvc

CZagrobelny committed:
remove unused image

CZagrobelny committed:
cleanup

CZagrobelny committed:
new line at end of file

CZagrobelny committed:
incorporate javier's feedback

Commit:
Add docker instructions to installfest

Commit:
Add docker instructions to Intro to Rails

Commit:
Adds docker instructions for Job Board

Commit:
Add docker glossary

Commit:
Add docker glossary

CZagrobelny committed:
fix syntax errors and split the create a rails app section into a seperate page for docker

CZagrobelny committed:
fix syntax and display issues

CZagrobelny committed:
formatting, fix syntax errors, add bundle install command for docker users

CZagrobelny committed:
Merge branch 'docker-intro-to-rails-edits' into may-2018-workshop

CZagrobelny committed:
Merge branch 'docker-job-board-edits' into may-2018-workshop

CZagrobelny committed:
Merge branch 'expand-mvc-focus-for-intro-to-rails' of github.com:railsbridge-nyc/docs into may-2018-workshop

CZagrobelny committed:
Merge branch 'add-hashes-to-ruby-basics' of github.com:railsbridge-nyc/docs into may-2018-workshop

CZagrobelny committed:
fix formatting for docker on the intro-to-rails step

CZagrobelny committed:
fix docker glossary formatting

CZagrobelny committed:
fix intro to rails docker glossary

javierjulio commented on sites/en/intro-to-rails/intro-to-rails.step:
```diff
@@ -38,31 +38,44 @@ day. Probably.
+    >> 1 + 2
+    => 3
+    >> require "active_support"
+    => true
```
We should remove the active_support portion here from the With Docker section as it won't work. It makes sense though for the Without Docker since users would have the gem installed after going through the InstallFest. I'm not sure if you prefer removing that from both or just leaving the math portion here?

CZagrobelny replied:
@javierjulio I think we can leave the math portion and remove active_support

javierjulio commented on sites/en/intro-to-rails/getting_started.step:
```diff
@@ -41,17 +41,95 @@ steps do
+
     message "`rails new` creates a new Rails project with the name you give."
     message "In this case we told it to create a new project called `suggestotron`. We'll go into detail on what it created shortly."
     message "This will print a lot of stuff to the screen and can take a while to finish."
```
We should copy these 3 messages into each of the 2 sub sections. 

It worked for the Without Docker section since its so small but the With Docker section is much longer, that it felt weird to read this after going through all that and waiting for the commands to finish.

1. I would use the existing text here for the Without Docker section. The way it is, is fine.
2. In the With Docker section I would suggest adjusting the text a little bit. Perhaps something like:

```
message "`rails new` creates a new Rails project with the name you give."
message "In this case we used `.` which is a shorthand for the current directory and which happens to be `suggestotron`. We'll go into detail on what it created shortly."
message "This will print a lot of stuff to the screen and can take a while to finish."
```

CZagrobelny replied:
@javierjulio these changes sound good to me!

javierjulio replied:
For the With Docker section I've changed the text slightly to now be:

> The `rails new` command creates a Rails project with the name you give. In this case we used `.` which is a shorthand for the current directory which happens to be `suggestotron`. We'll go into detail on what it created shortly. This and the following steps will all print a lot of stuff to the screen and might take a while to finish.

javierjulio commented on sites/en/intro-to-rails/running_your_application_locally.step:
```diff
@@ -20,19 +25,23 @@ every time you visit a page in your app."
   end
+  result <<-STOPPING_RAILS_SERVER
+    ^CExiting
+  STOPPING_RAILS_SERVER
```
Should we move this to the Without Docker section as it doesn't apply to With Docker? For Docker it would look something like the following after running `docker-compose down`:

```
Stopping suggestotron_web_1 ... done
Stopping suggestotron_db_1  ... done
Removing suggestotron_web_1     ... done
Removing suggestotron_web_run_3 ... done
Removing suggestotron_web_run_2 ... done
Removing suggestotron_web_run_1 ... done
Removing suggestotron_db_1      ... done
Removing network suggestotron_default
```

Do we want to include that or just remove the result portion? Perhaps just saying that now you should be able to type commands in terminal/console?

CZagrobelny replied:
@javierjulio maybe just remove the results section to keep it simple, but I am good with whatever you think is best.

javierjulio commented on sites/en/intro-to-rails/CRUD_with_scaffolding.step:
```diff
@@ -11,99 +10,57 @@ goals {
-    tip "Now is a good time to figure out how to have multiple tabs or windows of your terminal or command prompt. Starting and stopping the Rails server all day is tedious, so it's good to have one terminal tab or window for running commands, and a separate one for the server."
-  }
+  tip "Now is a good time to figure out how to have multiple tabs or windows of your terminal or command prompt. Starting and stopping the Rails server all day is tedious, so it's good to have one terminal tab or window for running commands, and a separate one for the server."
+  tip "Docker users should have a tab that is running rails server, `docker-compose up`, a tab that is running bash in the docker web service container, `docker-compose run web bash`, where you can run all your rails related commands and `irb`, and another tab to interact with git, heroku, or bringing down the server, `docker-compose down`."
```
The commas here threw me off a lot. While I was adjusting the wording I figured this would be better as a list and maybe easier to follow. This is what I came up with:

> Docker users should have a tab for each of the following:
> - `docker-compose up` for running the Rails server
> - `docker-compose run web bash` for the docker web service container, where you can run `irb` and all your Rails related commands
> - `docker-compose down` for shutting down the server or to interact with Git or Heroku

CZagrobelny replied:
@javierjulio this is great!  Much clearer!

javierjulio commented on sites/en/intro-to-rails/setting_the_default_page.step:
```diff
@@ -34,6 +34,7 @@ Rails.application.routes.draw do
 }
 
 explanation {
+  tip "Docker users should run rails commands in a terminal running `docker-compose run web bash`."
```
This is fine but I wondered if it could be improved a little at least to signify the steps taken in order to hopefully make it clearer.

> Docker users will need to run `docker-compose run web bash` first, where they can then run Rails related commands.

Not sure what you prefer but if we do change it seems this message is used on other pages too.

CZagrobelny replied:
@javierjulio if we have time, I think it's worth it to include the docker option every time we run a rails command...but I don't know how time consuming that will be to implement.

javierjulio reviewed (commented):
Some comments after I've gone through the InstallFest and up to the Allow People To Vote page in Intro To Rails.

javierjulio commented on sites/en/installfest/clean_up.step:
```diff
@@ -20,3 +20,7 @@ step "Delete the test_app from your computer" do
 end
+
+step "Delete the test_app Docker image from your computer" do
+  console "docker rmi testapp_web"
```
This didn't work but an easy fix. The correct name is **test_app_web**.

```
railsbridge: docker rmi testapp_web
Error: No such image: testapp_web
railsbridge: docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
test_app_web        latest              685701441146        About an hour ago   690MB
```

javierjulio reviewed (changes requested):
Correction for the Clean Up page as I was trying to remove the test app image.

javierjulio commented on sites/en/job-board/create_a_rails_app.step:
```diff
@@ -20,23 +20,108 @@ message <<-MARKDOWN
+    RUN mkdir -p $APP_HOME
+    WORKDIR $APP_HOME
+    COPY Gemfile* $APP_HOME/
+    RUN bundle install --binstubs
```
Remove the " --binstubs" portion otherwise after following the job_board instructions if you run a `bin/rails` command you get the following very verbose help output mixed with the output of the actual command you ran (e.g. generate a controller).

If I do the job_board setup with a Dockerfile that does NOT have the "--binstubs" portion I can then run any normal `rails` commands without the verbose binstub output pasted below. 

What I don't understand is why I didn't run into this on the IntroToRails setup. After that when I started the JobBoard I was running into this verbose binstub output.

```
job_board: (master) docker-compose run web rails generate controller jobs
Starting job_board_db_1 ... done
Beginning in Rails 4, Rails ships with a `rails` binstub at ./bin/rails that
should be used instead of the Bundler-generated `rails` binstub.

If you are seeing this message, your binstub at ./bin/rails was generated by
Bundler instead of Rails.

You might need to regenerate your `rails` binstub locally and add it to source
control:

 rails app:update:bin           # Bear in mind this generates other binstubs
                                # too that you may or may not want (like yarn)

If you already have Rails binstubs in source control, you might be
inadverently overwriting them during deployment by using bundle install
with the --binstubs option.

If your application was created prior to Rails 4, here's how to upgrade:

  bundle config --delete bin    # Turn off Bundler's stub generator
  rails app:update:bin          # Use the new Rails executables
  git add bin                   # Add bin/ to source control

You may need to remove bin/ from your .gitignore as well.

When you install a gem whose executable you want to use in your app,
generate it and add it to source control:

  bundle binstubs some-gem-name
  git add bin/new-executable

      create  app/controllers/jobs_controller.rb
      invoke  erb
      create    app/views/jobs
      invoke  helper
      create    app/helpers/jobs_helper.rb
      invoke  assets
      invoke    coffee
      create      app/assets/javascripts/jobs.coffee
      invoke    scss
      create      app/assets/stylesheets/jobs.scss
Usage:
  rails new APP_PATH [options]

Options:
      [--skip-namespace], [--no-skip-namespace]            # Skip namespace (affects only isolated applications)
  -r, [--ruby=PATH]                                        # Path to the Ruby binary of your choice
                                                           # Default: /usr/local/bin/ruby
  -m, [--template=TEMPLATE]                                # Path to some application template (can be a filesystem path or URL)
  -d, [--database=DATABASE]                                # Preconfigure for selected database (options: mysql/postgresql/sqlite3/oracle/frontbase/ibm_db/sqlserver/jdbcmysql/jdbcsqlite3/jdbcpostgresql/jdbc)
                                                           # Default: sqlite3
      [--skip-yarn], [--no-skip-yarn]                      # Don't use Yarn for managing JavaScript dependencies
      [--skip-gemfile], [--no-skip-gemfile]                # Don't create a Gemfile
  -G, [--skip-git], [--no-skip-git]                        # Skip .gitignore file
      [--skip-keeps], [--no-skip-keeps]                    # Skip source control .keep files
  -M, [--skip-action-mailer], [--no-skip-action-mailer]    # Skip Action Mailer files
  -O, [--skip-active-record], [--no-skip-active-record]    # Skip Active Record files
      [--skip-active-storage], [--no-skip-active-storage]  # Skip Active Storage files
  -P, [--skip-puma], [--no-skip-puma]                      # Skip Puma related files
  -C, [--skip-action-cable], [--no-skip-action-cable]      # Skip Action Cable files
  -S, [--skip-sprockets], [--no-skip-sprockets]            # Skip Sprockets files
      [--skip-spring], [--no-skip-spring]                  # Don't install Spring application preloader
      [--skip-listen], [--no-skip-listen]                  # Don't generate configuration that depends on the listen gem
      [--skip-coffee], [--no-skip-coffee]                  # Don't use CoffeeScript
  -J, [--skip-javascript], [--no-skip-javascript]          # Skip JavaScript files
      [--skip-turbolinks], [--no-skip-turbolinks]          # Skip turbolinks gem
  -T, [--skip-test], [--no-skip-test]                      # Skip test files
      [--skip-system-test], [--no-skip-system-test]        # Skip system test files
      [--skip-bootsnap], [--no-skip-bootsnap]              # Skip bootsnap gem
      [--dev], [--no-dev]                                  # Setup the application with Gemfile pointing to your Rails checkout
      [--edge], [--no-edge]                                # Setup the application with Gemfile pointing to Rails repository
      [--rc=RC]                                            # Path to file containing extra configuration options for rails command
      [--no-rc], [--no-no-rc]                              # Skip loading of extra configuration options from .railsrc file
      [--api], [--no-api]                                  # Preconfigure smaller stack for API only apps
  -B, [--skip-bundle], [--no-skip-bundle]                  # Don't run bundle install
      [--webpack=WEBPACK]                                  # Preconfigure for app-like JavaScript with Webpack (options: react/vue/angular/elm/stimulus)

Runtime options:
  -f, [--force]                    # Overwrite files that already exist
  -p, [--pretend], [--no-pretend]  # Run but do not make any changes
  -q, [--quiet], [--no-quiet]      # Suppress status output
  -s, [--skip], [--no-skip]        # Skip files that already exist

Rails options:
  -h, [--help], [--no-help]        # Show this help message and quit
  -v, [--version], [--no-version]  # Show Rails version number and quit

Description:
    The 'rails new' command creates a new Rails application with a default
    directory structure and configuration at the path you specify.

    You can specify extra command-line arguments to be used every time
    'rails new' runs in the .railsrc configuration file in your home directory.

    Note that the arguments specified in the .railsrc file don't affect the
    defaults values shown above in this help message.

Example:
    rails new ~/Code/Ruby/weblog

    This generates a skeletal Rails installation in ~/Code/Ruby/weblog.
```

CZagrobelny replied:
@javierjulio will defer to @maxjacobson and @pdarden since they are most familiar with " --binstubs"

maxjacobson replied:
I think we're safe to remove the `--binstubs` flag from the `Dockerfile`. That flag tells bundler to generate files under the `bin` directory so you can do things like `bin/rails` instead of `bundle exec rails`. AFAIK our docs don't encourage writing `bin/rails` or `bin/rake` at any point, so we're not dependent on those files existing

javierjulio replied:
We do have use of `bin/rails` in command instructions that we need to correct. Are we good with dropping the binstubs flag and then updating command examples to drop `bin/`? It would help to have someone follow through and confirm they get the same output as I pasted above following the JobBoard instructions.

maxjacobson replied:
Hm. Another option would be to change the line to:

```
RUN bundle install --binstubs && rails app:update:bin
```

That might be worthwhile as it's a smaller change, I think would avoid the (omg so long) warning and also not require changing other things. WDYT?

javierjulio replied:
@maxjacobson sure, that would be fine but would that really work? At that point there is no rails app so no binstub files to update. Or would it?

maxjacobson replied:
Good point, it might not work, on the first run, at which point rails is installed but the current working directory isn't (yet) a rails app. the `rails` binary might squawk about being in a non-rails-app context. It might be worth trying ...  if it doesn't work, then updating the references to `bin/rails` to be just `rails` (if that's a reasonably-small change) is probably the move

javierjulio committed:
Remove test active_support require

Doesn't make sense for Docker. The math portion is sufficient.

javierjulio committed:
Remove exit output section

This doesn't make sense to have with Docker. For now we'd rather remove it, although we could change this to move this into the Without Docker section and then in the With Docker section we could include the following exit output:

```
Stopping suggestotron_web_1 ... done
Stopping suggestotron_db_1  ... done
Removing suggestotron_web_1     ... done
Removing suggestotron_web_run_3 ... done
Removing suggestotron_web_run_2 ... done
Removing suggestotron_web_run_1 ... done
Removing suggestotron_db_1      ... done
Removing network suggestotron_default
```

javierjulio committed:
Fix app name in docker command

javierjulio committed:
Adjust wording to reflect order of operations

javierjulio committed:
Breakdown docker comands for multiple tabs in a list

javierjulio committed:
Copy the rails new info to both With/Without sections

Since the With Docker section is much longer due to the content that has to be copied into files, it felt really weird to be reading about `rails new` and that it would take awhile to run and output a lot of info. I included this with slightly different wording for the With Docker section since it uses `.` as the rails app name which means the current directory.

javierjulio committed:
Update Docker Glossary

While we describe each command, they are not always helpful in the context of the workshop, so for example when describing `docker-compose down` we now mention that if you have a Rails server running, this is the command you use to stop it.

javierjulio commented:
@CZagrobelny @maxjacobson @pdarden the one thing remaining is I had updated one of the tips to now be:

> tip "Docker users will need to run `docker-compose run web bash` first, where they can then run Rails related commands."

which will likely be helpful to have in more pages. Should we include that or what we have for now is fine? If someone could help me identify some good spots to place it, I can copy it there.

javierjulio committed:
Remove `--binstubs` from install and any user use of `bin/rails`

We need to remove the " --binstubs" portion otherwise after following the job_board instructions if you run a `bin/rails` command you get the following very verbose help output **mixed with the output of the actual command you ran** (e.g. generate a controller) so that info gets lost in the mix.

If I do the job_board setup with a Dockerfile that does NOT have the "--binstubs" portion I can then run any normal `rails` commands without the verbose binstub output pasted below.

What I don't understand is why I didn't run into this on the IntroToRails setup. After that when I started the JobBoard I was running into this verbose binstub output.

```
job_board: (master) docker-compose run web rails generate controller jobs
Starting job_board_db_1 ... done
Beginning in Rails 4, Rails ships with a `rails` binstub at ./bin/rails that
should be used instead of the Bundler-generated `rails` binstub.

If you are seeing this message, your binstub at ./bin/rails was generated by
Bundler instead of Rails.

You might need to regenerate your `rails` binstub locally and add it to source
control:

 rails app:update:bin           # Bear in mind this generates other binstubs
                                # too that you may or may not want (like yarn)

If you already have Rails binstubs in source control, you might be
inadverently overwriting them during deployment by using bundle install
with the --binstubs option.

If your application was created prior to Rails 4, here's how to upgrade:

  bundle config --delete bin    # Turn off Bundler's stub generator
  rails app:update:bin          # Use the new Rails executables
  git add bin                   # Add bin/ to source control

You may need to remove bin/ from your .gitignore as well.

When you install a gem whose executable you want to use in your app,
generate it and add it to source control:

  bundle binstubs some-gem-name
  git add bin/new-executable

      create  app/controllers/jobs_controller.rb
      invoke  erb
      create    app/views/jobs
      invoke  helper
      create    app/helpers/jobs_helper.rb
      invoke  assets
      invoke    coffee
      create      app/assets/javascripts/jobs.coffee
      invoke    scss
      create      app/assets/stylesheets/jobs.scss
Usage:
  rails new APP_PATH [options]

Options:
      [--skip-namespace], [--no-skip-namespace]            # Skip namespace (affects only isolated applications)
  -r, [--ruby=PATH]                                        # Path to the Ruby binary of your choice
                                                           # Default: /usr/local/bin/ruby
  -m, [--template=TEMPLATE]                                # Path to some application template (can be a filesystem path or URL)
  -d, [--database=DATABASE]                                # Preconfigure for selected database (options: mysql/postgresql/sqlite3/oracle/frontbase/ibm_db/sqlserver/jdbcmysql/jdbcsqlite3/jdbcpostgresql/jdbc)
                                                           # Default: sqlite3
      [--skip-yarn], [--no-skip-yarn]                      # Don't use Yarn for managing JavaScript dependencies
      [--skip-gemfile], [--no-skip-gemfile]                # Don't create a Gemfile
  -G, [--skip-git], [--no-skip-git]                        # Skip .gitignore file
      [--skip-keeps], [--no-skip-keeps]                    # Skip source control .keep files
  -M, [--skip-action-mailer], [--no-skip-action-mailer]    # Skip Action Mailer files
  -O, [--skip-active-record], [--no-skip-active-record]    # Skip Active Record files
      [--skip-active-storage], [--no-skip-active-storage]  # Skip Active Storage files
  -P, [--skip-puma], [--no-skip-puma]                      # Skip Puma related files
  -C, [--skip-action-cable], [--no-skip-action-cable]      # Skip Action Cable files
  -S, [--skip-sprockets], [--no-skip-sprockets]            # Skip Sprockets files
      [--skip-spring], [--no-skip-spring]                  # Don't install Spring application preloader
      [--skip-listen], [--no-skip-listen]                  # Don't generate configuration that depends on the listen gem
      [--skip-coffee], [--no-skip-coffee]                  # Don't use CoffeeScript
  -J, [--skip-javascript], [--no-skip-javascript]          # Skip JavaScript files
      [--skip-turbolinks], [--no-skip-turbolinks]          # Skip turbolinks gem
  -T, [--skip-test], [--no-skip-test]                      # Skip test files
      [--skip-system-test], [--no-skip-system-test]        # Skip system test files
      [--skip-bootsnap], [--no-skip-bootsnap]              # Skip bootsnap gem
      [--dev], [--no-dev]                                  # Setup the application with Gemfile pointing to your Rails checkout
      [--edge], [--no-edge]                                # Setup the application with Gemfile pointing to Rails repository
      [--rc=RC]                                            # Path to file containing extra configuration options for rails command
      [--no-rc], [--no-no-rc]                              # Skip loading of extra configuration options from .railsrc file
      [--api], [--no-api]                                  # Preconfigure smaller stack for API only apps
  -B, [--skip-bundle], [--no-skip-bundle]                  # Don't run bundle install
      [--webpack=WEBPACK]                                  # Preconfigure for app-like JavaScript with Webpack (options: react/vue/angular/elm/stimulus)

Runtime options:
  -f, [--force]                    # Overwrite files that already exist
  -p, [--pretend], [--no-pretend]  # Run but do not make any changes
  -q, [--quiet], [--no-quiet]      # Suppress status output
  -s, [--skip], [--no-skip]        # Skip files that already exist

Rails options:
  -h, [--help], [--no-help]        # Show this help message and quit
  -v, [--version], [--no-version]  # Show Rails version number and quit

Description:
    The 'rails new' command creates a new Rails application with a default
    directory structure and configuration at the path you specify.

    You can specify extra command-line arguments to be used every time
    'rails new' runs in the .railsrc configuration file in your home directory.

    Note that the arguments specified in the .railsrc file don't affect the
    defaults values shown above in this help message.

Example:
    rails new ~/Code/Ruby/weblog

    This generates a skeletal Rails installation in ~/Code/Ruby/weblog.
```

CZagrobelny committed:
dont use -b 0.0.0.0 when starting rails in the non-docker option
